redis_commands.c

```diff
@@ -728,6 +736,148 @@ int redis_zrangebyscore_cmd(INTERNAL_FUNCTION_PARAMETERS, RedisSock *redis_sock,
+        return FAILURE;
+    } else if (Z_TYPE_P(val) == IS_ARRAY && zend_hash_num_elements(Z_ARRVAL_P(val)) == 0) {
+        php_error_docref(NULL, E_WARNING, "Cannot pass an empty array to CONFIG GET");
+        return FAILURE;
+    }
+
+    ncfg = Z_TYPE_P(val) == IS_STRING ? 1 : zend_hash_num_elements(Z_ARRVAL_P(val));
+
+    REDIS_CMD_INIT_SSTR_STATIC(dst, 1 + ncfg, "CONFIG");
+    REDIS_CMD_APPEND_SSTR_STATIC(dst, "GET");
+
+    if (Z_TYPE_P(val) == IS_STRING) {
+        redis_cmd_append_sstr_zstr(dst, Z_STR_P(val));
+    } else {
+        ZEND_HASH_FOREACH_VAL(Z_ARRVAL_P(val), zv) {
+            ZVAL_DEREF(zv);
+
+            zstr = zval_get_string(zv);
+            redis_cmd_append_sstr_zstr(dst, zstr);
+            zend_string_release(zstr);
+        } ZEND_HASH_FOREACH_END();
+    }
+
+    return SUCCESS;
+}
+
+static int redis_build_config_set_cmd(smart_string *dst, zval *key, zend_string *val) {
+    zend_string *zkey, *zstr;
+    zval *zv;
+
+    /* Legacy case:  CONFIG SET <string> <string> */
+    if (key != NULL && val != NULL) {
+        REDIS_CMD_INIT_SSTR_STATIC(dst, 3, "CONFIG");
+        REDIS_CMD_APPEND_SSTR_STATIC(dst, "SET");
+
+        zstr = zval_get_string(key);
+        redis_cmd_append_sstr_zstr(dst, zstr);
+        zend_string_release(zstr);
+
+        redis_cmd_append_sstr_zstr(dst, val);
+
+        return SUCCESS;
+    }
+
+    /* Now we must have an array with at least one element */
+    if (key == NULL || Z_TYPE_P(key) != IS_ARRAY || zend_hash_num_elements(Z_ARRVAL_P(key)) == 0) {
+        php_error_docref(NULL, E_WARNING, "Must either pass two strings to CONFIG SET or a non-empty array of values");
+        return FAILURE;
+    }
+
+    REDIS_CMD_INIT_SSTR_STATIC(dst, 1 + (2 * zend_hash_num_elements(Z_ARRVAL_P(key))), "CONFIG");
+    REDIS_CMD_APPEND_SSTR_STATIC(dst, "SET");
+
+    ZEND_HASH_FOREACH_STR_KEY_VAL(Z_ARRVAL_P(key), zkey, zv) {
+        if (zkey == NULL)
+            goto fail;
+
+        ZVAL_DEREF(zv);
+
+        redis_cmd_append_sstr_zstr(dst, zkey);
+
+        zstr = zval_get_string(zv);
+        redis_cmd_append_sstr_zstr(dst, zstr);
+        zend_string_release(zstr);
+    } ZEND_HASH_FOREACH_END();
+
+    return SUCCESS;
+
+fail:
+    php_error_docref(NULL, E_WARNING, "Must pass an associate array of config keys and values");
+    efree(dst->c);
+    memset(dst, 0, sizeof(*dst));
+    return FAILURE;
+}
+
+int
+redis_config_cmd(INTERNAL_FUNCTION_PARAMETERS, RedisSock *redis_sock,
+                 char **cmd, int *cmd_len, short *slot, void **ctx)
+{
+    zend_string *op = NULL, *arg = NULL;
+    smart_string cmdstr = {0};
+    enum redisConfigOp cfg_op;
+    int res = FAILURE;
+    zval *key = NULL;
+
+    ZEND_PARSE_PARAMETERS_START(1, 3)
+        Z_PARAM_STR(op)
+        Z_PARAM_OPTIONAL
+        Z_PARAM_ZVAL_OR_NULL(key)
+        Z_PARAM_STR_OR_NULL(arg)
+    ZEND_PARSE_PARAMETERS_END_EX(return FAILURE);
+
+    if (redis_get_config_op(&cfg_op, op) != SUCCESS) {
```

Comment:
I think you can just do checks and build command here and not create additional enum+ switch

Reply:
Yep, way simpler, good catch!